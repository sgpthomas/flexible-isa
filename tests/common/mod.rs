use std::{
    fmt::Display,
    path::{Path, PathBuf},
};

use flexible_isa::{cli, run, HalideExprOp, InstructionSelect};
use itertools::Itertools;

const BENCHMARKS_PATH: &str = "cache";
#[allow(unused)]
const BENCHMAKRS: &[&str] = &[
    "add",
    "average_pool",
    "blur3x3",
    "camera_pipe",
    "conv3x3a16",
    "conv3x3a32",
    "conv_nn",
    "depthwise_conv",
    "dilate3x3",
    // "elementwise",
    // "fully_connected",
    "gaussian3x3",
    "gaussian5x5",
    "gaussian7x7",
    "l2norm",
    "matmul",
    "max_pool",
    "mean",
    "median3x3",
    "mul",
    "sobel3x3",
    "softmax",
];

pub fn test_single_isa<S: AsRef<Path> + Display>(name: S) -> anyhow::Result<()> {
    let mut path = PathBuf::from(BENCHMARKS_PATH).join(&name);
    path.set_extension("stmt");

    let args: cli::Args = cli::Args::new(&[path]).learn(true);
    let isa = run(args)?;

    let instr_hist = InstructionSelect::from_recexpr(&isa.expressions);
    let hist: Vec<_> = instr_hist
        .iter()
        .filter_map(|(op, count)| {
            if let HalideExprOp::Instruction(i) = op {
                Some((i, count))
            } else {
                None
            }
        })
        // sort first by count, then by index
        .sorted_by_key(|(i, count)| (*count, *i))
        .map(|(i, count)| (i, count, isa.instructions[i].to_string()))
        .collect();

    insta::with_settings!({
        description => format!("generate isa for {name}"),
        omit_expression => true,
        snapshot_suffix => format!("{name}")
    }, {
        insta::assert_debug_snapshot!(hist)
    });

    Ok(())
}
