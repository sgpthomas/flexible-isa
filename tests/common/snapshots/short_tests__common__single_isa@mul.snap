---
source: tests/common/mod.rs
description: generate isa for mul
---
(
    MinimalIsa {
      10: (+ (cast<int32> ?x0) (cast<int32> ?x1))
      21: (- ?x0 ?x1)
      27: (* ?x0 ?x1)
      28: (/ (cast<int32> ?#0) (cast<int32> ?#1))
      29: (% (cast<int32> ?#0) (cast<int32> ?#1))
      30: (<= ?x0 ?x1)
      31: (== (cast<int32> ?#0) (cast<int32> ?#1))
      32: (!= (cast<int32> ?#0) (cast<int32> ?#1))
      34: (get
            ?x0
            (cast<int32x128>
              (call[ramp]
                (cast<int32>
                  (+
                    (cast<int32>
                      (* (cast<int32> ?x1) (cast<int32> (+ (cast<int32> ?#7) (cast<int32> ?#8)))))
                    (cast<int32>
                      (-
                        (cast<int32>
                          (-
                            (cast<int32>
                              (+
                                (cast<int32>
                                  (* (cast<int32> (/ (cast<int32> ?#9) (cast<int32> 128))) (cast<int32> 128)))
                                (cast<int32> ?#10)))
                            (cast<int32> (* (cast<int32> ?x2) (cast<int32> ?x1)))))
                        (cast<int32> ?x3)))))
                (cast<int32> 1)
                (cast<int32> 128)))
            (cast<bool> (<= ?x4 ?x5)))
      37: (get ?x0 ?x1)
    }
    ,
    [
        (
            i29,
            2,
            "(\"%\" (cast<int32> ?#0) (cast<int32> ?#1))",
        ),
        (
            i30,
            2,
            "(<= ?x0 ?x1)",
        ),
        (
            i31,
            2,
            "(== (cast<int32> ?#0) (cast<int32> ?#1))",
        ),
        (
            i32,
            2,
            "(!= (cast<int32> ?#0) (cast<int32> ?#1))",
        ),
        (
            i34,
            3,
            "(get ?x0 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x1) (cast<int32> (+ (cast<int32> ?#7) (cast<int32> ?#8))))) (cast<int32> (- (cast<int32> (- (cast<int32> (+ (cast<int32> (* (cast<int32> (/ (cast<int32> ?#9) (cast<int32> 128))) (cast<int32> 128))) (cast<int32> ?#10))) (cast<int32> (* (cast<int32> ?x2) (cast<int32> ?x1))))) (cast<int32> ?x3))))) (cast<int32> 1) (cast<int32> 128))) (cast<bool> (<= ?x4 ?x5)))",
        ),
        (
            i37,
            5,
            "(get ?x0 ?x1)",
        ),
        (
            i28,
            9,
            "(/ (cast<int32> ?#0) (cast<int32> ?#1))",
        ),
        (
            i21,
            22,
            "(- ?x0 ?x1)",
        ),
        (
            i27,
            36,
            "(* ?x0 ?x1)",
        ),
        (
            i10,
            38,
            "(+ (cast<int32> ?x0) (cast<int32> ?x1))",
        ),
    ],
)
